### PR TITLE
Initial Social Login implementation

### PR DESCRIPTION
**Pull request by hcho112**
This PR contains initial social login of google and apple (facebook is WIP) 
Main difference compared to existing solution is that:
1. We use social login of selected platform to either create or sign in
2. Even user starts at sign in, if near account doesn't exist, it take create flow

You can review created account on firebase console. (To test creation after you already used your own social login, it would be easier to delete user record on firebase console and try to create again)

Currently I use random string to generate account Id. (This idea was originated from other platform that also uses social login, they also use random string to auto-generate their account id of their user. Intention is to remove extra steps of keying in account id. Assuming that user will always use social login, this is not a bad idea, however I'm happy to replace this after showcase it to rest of the team.) 

This mode is currently only available on testnet. 

On mainnet:
Greenshot 2024-07-19 16 35 24
Greenshot 2024-07-19 16 35 11

On testnet:
Greenshot 2024-07-19 16 14 06
Greenshot 2024-07-19 16 02 41

### CODE REVIEW

File: packages/near-fast-auth-signer/src/components/SocialLogin/SocialButton.tsx
Comment (Pessina) [inline packages/near-fast-auth-signer/src/components/SocialLogin/SocialButton.tsx:73]: Using then/catch, async/await. Isn't better choose one and go with it?

 I find async/wait more clear

File: packages/near-fast-auth-signer/src/components/SocialLogin/SocialButton.tsx
Comment (Pessina) [inline packages/near-fast-auth-signer/src/components/SocialLogin/SocialButton.tsx:89]: I see the code here share a lot of similarities with AuthCallback page. Can we create functions and re-use them to avoid maintaining the same code on two parts? 

https://github.com/near/fast-auth-signer/blob/8d2726f5fe007d7ae011dec245692ac7499aff0c/packages/near-fast-auth-signer/src/components/AuthCallback/AuthCallback.tsx#L238

File: packages/near-fast-auth-signer/src/utils/string.ts
Comment (Pessina) [inline packages/near-fast-auth-signer/src/utils/string.ts:1]: I think we have the UUID library installed for this. 

It's also better for collision proof

**Review by Pessina — CHANGES_REQUESTED** [review]

**Comment by frol** [comment]
@hcho112 @Pessina Sorry for jumping in, but I'd like to invite you to participate in the [Race of Sloths](https://race-of-sloths.com) - a fun and gamified open-source contributions program. Consider mentioning @race-of-sloths user in your github comment or PR description to join!

See how the flow works here: https://github.com/near/nearcore/pull/11778

**Comment by hcho112** [comment]
Currently there is an issue with e2e test failures due to smart contract associated on multichain tests. It has nothing to do with this PR. The team is working on resolving the issue

**Comment by hcho112** [comment]
@race-of-sloths

**Comment by race-of-sloths** [comment]
@hcho112 Thank you for your contribution! Your pull request is now a part of the [Race of Sloths](https://race-of-sloths.com)!
New Sloth joined the Race! Welcome!

[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://badge.race-of-sloths.com/hcho112?type=bot&theme=dark">
    <source media="(prefers-color-scheme: light)" srcset="https://badge.race-of-sloths.com/hcho112?type=bot&theme=light">
    Shows profile picture for the author of the PR
</picture>](https://race-of-sloths.com/profile/hcho112)


<details><summary>Current status: <i>waiting for scoring</i></summary>

We're waiting for maintainer to score this pull request with `@race-of-sloths score [0,1,2,3,5,8,13]` command
</details><details><summary>What is the Race of Sloths</summary>
<p>
Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag @race-of-sloths inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/hcho112)

For maintainers:
- Score pull requests that participate in the Race of Sloths
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!
</p>
</details> 

<details>
<summary>Bot commands</summary>

- For contributors
  - **Include a PR:** `@race-of-sloths include` to enter the Race with your PR
- For maintainers:
  - **Assign points:** `@race-of-sloths score [1/2/3/5/8/13]` to award points based on your assessment.
  - **Reject this PR:** `@race-of-sloths exclude` to send this PR back to the drawing board.
  - **Exclude repo:** `@race-of-sloths pause` to stop bot activity in this repo until `@race-of-sloths unpause` command is called

</details>

File: packages/near-fast-auth-signer/src/components/SocialLogin/SocialButton.tsx
Comment (Pessina) [inline packages/near-fast-auth-signer/src/components/SocialLogin/SocialButton.tsx:94]: This still duplicated no? 

https://github.com/near/fast-auth-signer/blob/8d2726f5fe007d7ae011dec245692ac7499aff0c/packages/near-fast-auth-signer/src/components/AuthCallback/AuthCallback.tsx#L248

I think it would be good to extract the common logic from AuthCallback and SocialLogin to a separated component

1 - passkeys
2 - claimoidc
3 - call createAccount or signIn

What do you think? Otherwise we have to maintain the same flow on two places.

Comment (benmalcom) [reply packages/near-fast-auth-signer/src/components/SocialLogin/SocialButton.tsx:94]: Also I've extracted some function in the AuthCallback component in my PR, I think we should look for how to reconcile them.

Comment (hcho112) [reply packages/near-fast-auth-signer/src/components/SocialLogin/SocialButton.tsx:94]: I thought previous comment was about refactoring the part that I pull out and replace with `setAccountIdToController`. 

I did notice that @benmalcom also made refactor on this PR, perhaps I will reuse his one. (Either we merge this PR first and once @benmalcom 's [PR](https://github.com/near/fast-auth-signer/pull/264) merged, I refactor or we merge his one first and I will rebase on top of his)

Also the this part is actually different to Authcallback

```
const isNewUser = shouldCreateAccount(user.metadata.creationTime, user.metadata.lastSignInTime);
      let callback;
      if (isNewUser) {
        callback = onCreateAccount;
      } else if (isRecovery) {
        callback = onSignIn;
      } else {
        callback = onCreateAccount;
      }

```

so its not fully refactor-able

Comment (Pessina) [reply packages/near-fast-auth-signer/src/components/SocialLogin/SocialButton.tsx:94]: Cool! No problem then, let's work on it after merge both PRs.

File: packages/near-fast-auth-signer/src/components/SocialLogin/SocialButton.tsx
Comment (Pessina) [inline packages/near-fast-auth-signer/src/components/SocialLogin/SocialButton.tsx:106]: Can you use the `https://firebase.google.com/docs/reference/js/auth.md#operationtype` available on the `result` from ` const result = await signInWithPopup(firebaseAuth, provider);`? Instead of the custom function to check if it's account creation?

File: packages/near-fast-auth-signer/src/components/SocialLogin/SocialButton.tsx
Comment (Pessina) [inline packages/near-fast-auth-signer/src/components/SocialLogin/SocialButton.tsx:69]: Either way is fine, just sharing knowledge, but you can type a component with:

`const SocialButton: React.FC<SocialButtonProps> = (props) => {....}`

So we can improve a little bit the type of the function.

**Review by Pessina — CHANGES_REQUESTED** [review]

File: packages/near-fast-auth-signer/src/components/SocialLogin/SocialButton.tsx
Comment (hcho112) [reply packages/near-fast-auth-signer/src/components/SocialLogin/SocialButton.tsx:106]: Unfortunately, it will always return with `signIn` which it cant used to distinguish if we are using it for create or sign in

File: packages/near-fast-auth-signer/src/components/AuthCallback/auth.ts
Comment (Pessina) [inline packages/near-fast-auth-signer/src/components/AuthCallback/auth.ts:188]: Instead of string you can specify the strings allowed here.

**Review by Pessina — APPROVED** [review]

**Review by benmalcom — APPROVED** [review]
Looks good.

**Comment by race-of-sloths** [comment]
🥁  __Score it!__

@Pessina @benmalcom, please score the PR with `@race-of-sloths score [1/2/3/5/8/13]`. The contributor deserves it.
If no scoring is provided within 24 hours, this PR will be scored as 2 🦥

**Pull request merged**
